### PR TITLE
[bug] hdf5/1.12.1 (any version) : Include the module_file in cmake_build_modules for targets

### PR DESCRIPTION
**Pull request by paulharris**
Specify library name and version:  **hdf5/1.12.1**

The recipe created a cmake module but did not include it in the "cmake_build_modules" set.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/hdf5/all/conanfile.py [hunk 1]
```diff
@@ -229,6 +229,7 @@ def _config_libname(lib):
 
             self.cpp_info.components[component_name].set_property("cmake_target_name", f"hdf5::{alias_target}")
             self.cpp_info.components[component_name].set_property("pkg_config_name", alias_target)
+            self.cpp_info.components[component_name].set_property("cmake_build_modules", [self._module_file_rel_path])
```
Comment (SpaceIm) [inline recipes/hdf5/all/conanfile.py:232]:
```suggestion
```

no, CMakeDeps doesn't need this module file, it's a workaround for cmake_find_package

Comment (paulharris) [reply recipes/hdf5/all/conanfile.py:232]: I've had a go, please check

File: recipes/hdf5/all/conanfile.py [hunk 2]
```diff
@@ -193,14 +193,23 @@ def _create_cmake_module_alias_targets(module_file, targets, is_parallel):
                     set_property(TARGET hdf5::hdf5_hl_cpp PROPERTY INTERFACE_LINK_LIBRARIES HDF5::HL_CXX)
                 endif()
             """)
-        content += textwrap.dedent("set(HDF5_IS_PARALLEL {})".format("ON" if is_parallel else "OFF"))
+        tools.save(module_file, content)
+
+    @staticmethod
+    def _create_cmake_module_variables(module_file, is_parallel):
+        content = textwrap.dedent("set(HDF5_IS_PARALLEL {})".format("ON" if is_parallel else "OFF"))
```
Comment (SpaceIm) [inline recipes/hdf5/all/conanfile.py:200]:
```suggestion
        content = "set(HDF5_IS_PARALLEL {})".format("ON" if is_parallel else "OFF")
```

Comment (paulharris) [reply recipes/hdf5/all/conanfile.py]: Thanks!

**Review by SpaceIm — APPROVED** [review]

**Comment by paulharris** [comment]
Don't merge yet,
I'm looking a little deeper into hdf5's link-to-zlib behaviour, related to package: cgns

**Comment by uilianries** [comment]
> Don't merge yet,

Convert it to a draft! GH will not allow to be merged.

**Comment by paulharris** [comment]
Seems to be ok now.  Note that its "parallel" feature didn't work prior to this patch, and I wasn't able to fix it.
I see an open PR for openmpi from SpaceIM, maybe that will fix it.

**Comment by paulharris** [comment]
@uilianries can we please merge this?

the "parallel" problem mentioned before is a separate problem that should not block this merge.
#10479 may go some way to resolving it,
but I think the problem is the OpenMPI recipe doesn't export all of the cmake variables that the standard FindMPI.cmake exports, https://cmake.org/cmake/help/latest/module/FindMPI.html
And that fix is beyond me at the moment.

Note the parallel problem existed before this pull, and isn't a flag set by default.

File: recipes/hdf5/all/conanfile.py [hunk 3]
```diff
@@ -229,14 +241,16 @@ def _config_libname(lib):
 
             self.cpp_info.components[component_name].set_property("cmake_target_name", f"hdf5::{alias_target}")
             self.cpp_info.components[component_name].set_property("pkg_config_name", alias_target)
+            self.cpp_info.components[component_name].set_property("cmake_build_modules", [self._module_file_variables_rel_path])
             self.cpp_info.components[component_name].libs = [_config_libname(alias_target)]
             self.cpp_info.components[component_name].requires = requirements
+            self.cpp_info.components[component_name].builddirs = [os.path.join("lib", "cmake")]
```
Comment (SpaceIm) [inline recipes/hdf5/all/conanfile.py:247]:
```suggestion
```

Comment (SpaceIm) [reply recipes/hdf5/all/conanfile.py:247]: hook is wrong, it's useless.

**Comment by uilianries** [comment]
@paulharris There is a rule which must be followed to merge any PR: https://github.com/conan-io/conan-center-index/blob/master/docs/review_process.md

Your PR must pass to be merged. No exceptions.

File: recipes/hdf5/all/conanfile.py (hunk 3 above)
Comment (paulharris) [reply recipes/hdf5/all/conanfile.py]: Ok thanks @SpaceIm ,
please see latest commit to change it to append,
however, when you say the "hook is useless", do you mean I should not be appending to builddirs?
Or is this ok now?

**Comment by paulharris** [comment]
> @paulharris There is a rule which must be followed to merge any PR: https://github.com/conan-io/conan-center-index/blob/master/docs/review_process.md
> 
> Your PR must pass to be merged. No exceptions.

Ah sorry I didn't see the failure for some reason.  I thought it built ok.

File: recipes/hdf5/all/conanfile.py (hunk 3 above)
Comment (SpaceIm) [reply recipes/hdf5/all/conanfile.py]: You should not append nor override here, the hook is wrong.

Comment (paulharris) [reply recipes/hdf5/all/conanfile.py]: I do have a 2 cmake_build_modules file that is generated ...
ah maybe because I didn't include it in that set_property("cmake_build_modules") line 245...

**Comment by paulharris** [comment]
All the linux variants built with shared/unshared, until it got to GCC 11 (shared).
It wasn't able to link to zlib, but this is a shared lib, meaning for conan, it builds with the static zlib.
So I'm not sure what could be the problem or the fix ...

**Review by SSE4 — APPROVED** [review]

**Comment by paulharris** [comment]
@SSE4 I'm at a complete loss of why the build is failing ... help me Obi-wan ...

**Comment by SSE4** [comment]
these are `zlib` symbols:
```
Undefined symbols for architecture x86_64:
  "_compress2", referenced from:
      _H5Z_filter_deflate in H5Zdeflate.c.o
  "_inflate", referenced from:
      _H5Z_filter_deflate in H5Zdeflate.c.o
  "_inflateEnd", referenced from:
      _H5Z_filter_deflate in H5Zdeflate.c.o
  "_inflateInit_", referenced from:
      _H5Z_filter_deflate in H5Zdeflate.c.o
```
it seems like zlib wasn't properly linked in? 🤔

**Comment by SSE4** [comment]
> How did it pass before?

I am not entirely sure (it's not that trivial to check), but I think it entered a new "epoch" this time - as it builds for the compiler that was added recently enough to CCI : `compiler.version=13.0`. I believe older version was built only up to 12 or so?

**Comment by paulharris** [comment]
No it was failing for compiler 11, shared=True.
And succeed for compiler 10, shared=True and every other linux.
I'll dig up a link from further back here.

But please check #10983 for more.  Nothing to do with new compilers...

**Comment by paulharris** [comment]
Here is an example where everything built until it hit GCC 11.
https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/10836/6-linux-gcc/hdf5/1.12.1//summary.json

The problem is they aren't linking zlib.  It is the CMakeLists.txt fault, it has to be patched to work.
However, I don't understand how GCC 5,7,8,9,10 was able to build.


Nor do I understand how it was able to build in #10552 
https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/10552/2/hdf5/1.12.1//summary.json
... same compiler, same issues (recipe did not patch zlib), but, it linked just fine.

**Comment by paulharris** [comment]
I expect this PR will go green and can be merged.
But, it was a huge time suck due to the confusing CI behaviour.
I would really like to have the problems followed up in PR #10983 because I would not want to go through that again.

**Comment by uilianries** [comment]
> Here is an example where everything built until it hit GCC 11.
[c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/10836/6-linux-gcc/hdf5/1.12.1//summary.json](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/10836/6-linux-gcc/hdf5/1.12.1//summary.json)

We use Docker images for Linux builds. From GCC5 to GCC10 we use what we call "legacy". GCC 11 uses a new flavor of image based on Ubuntu 16.04, which contains a middle term of version of glibc (due compatibility, etc ...) 

> The problem is they aren't linking zlib. It is the CMakeLists.txt fault, it has to be patched to work.
However, I don't understand how GCC 5,7,8,9,10 was able to build.

In GCC11 image, we install and run CMake from Conan Center. All other versions are running CMake built from sources in the image.

**Comment by paulharris** [comment]
> > Here is an example where everything built until it hit GCC 11.
> > [c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/10836/6-linux-gcc/hdf5/1.12.1//summary.json](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/10836/6-linux-gcc/hdf5/1.12.1//summary.json)
> 
> We use Docker images for Linux builds. From GCC5 to GCC10 we use what we call "legacy". GCC 11 uses a new flavor of image based on Ubuntu 16.04, which contains a middle term of version of glibc (due compatibility, etc ...)
> 
> > The problem is they aren't linking zlib. It is the CMakeLists.txt fault, it has to be patched to work.
> > However, I don't understand how GCC 5,7,8,9,10 was able to build.
> 
> In GCC11 image, we install and run CMake from Conan Center. All other versions are running CMake built from sources in the image.

Would it be possible to provide a docker-run command to reproduce what is happening on the CI ?
I tried to use the conan docker images, however it built just fine on that docker,
so it is really hard to reproduce and diagnose problems like that.

Also, I know we can enable verbose makefiles and enable "trace" logging for cmake,
but is it possible to enable --trace-expand on the cmake calls to help diagnose what is going on?

Note that I've done a final clean-up of the recipe, so when it has finished building (again) we should be good to merge.

**Comment by paulharris** [comment]
@uilianries I was not able to duplicate the CI error on my own docker.

I was able to build the same Package_ID that failed, above, the link: https://github.com/conan-io/conan-center-index/pull/10836#issuecomment-1141784653
```
    Package_ID: ffff4d4c91914df27e9be9afe30af3d20e8f424b
        [options]
            enable_cxx: True
            hl: True
            parallel: False
            shared: True
            szip_support: None
            with_zlib: True
        [settings]
            arch: x86_64
            build_type: Release
            compiler: clang
            compiler.libcxx: libstdc++
            compiler.version: 12
            os: Linux
        [requires]
            zlib/1.2.12:cf142e38e022cf2a99aadebfd8f37ead3e935ad7
        Outdated from recipe: False
```

Using the commands:
```
cd /tmp
git clone https://github.com/paulharris/conan-center-index
cd conan-center-index/recipes/hdf5
git checkout -b bug-hdf5-cmake-build-modules origin/bug-hdf5-cmake-build-modules
docker run --network=host --rm -ti -v ${PWD}:/home/conan/project conanio/clang12-ubuntu16.04:1.47.0 

DOCKER>
cd project/all
conan create . 1.10.5@ --build=missing -o "hdf5:shared=True"
conan test test_package/conanfile.py hdf5/1.10.5@ -o "hdf5:shared=True"
```

This all worked successfully, unlike the CI ...

What am I missing?

**Comment by paulharris** [comment]
I'm going to have to give up on this PR.
If the CI has a problem, and I can't reproduce it in a place where I can diagnose the problem, then it goes into the too-hard basket eventually.

I'll probably just add the problematic platforms to the "incompatible configuration" list so it will build in the CI, and move on.

**Comment by jgsogo** [comment]
Please, @jcar87 . Can you have a look at this PR and the underlying issue? Thanks!

**Comment by jcar87** [comment]
Thanks @paulharris for opening a PR to fix up the handling of the CMake module files.

I have looked into a the spurious build/test failures and I have a theory about the issues linking ZLIB. However from what I can see, this PR as it is is correct and should not fail, and the current spurious failures are caused by an HTTP timeout retrieving the sources, rather than related to the recipe or Conan Center CI.

### ZLIB linking errors

The logic to link to ZLIB in the upstream build scripts is a bit fragile: https://github.com/HDFGroup/hdf5/blob/develop/CMakeFilters.cmake#L63-L67

According to the above, ZLIB will be linked correctly if the following 3 variables area defined: `ZLIB_FOUND`, `ZLIB_LIBRARIES` and `ZLIB_INCLUDE_DIRS`.

Up until this PR, CMake's own `FindZLIB.cmake` was used to locate zlib (which it locates successfully from the conan cache) - no problems there. 

If there is a situation where `ZLIB_FOUND` but the other two variables are not defined, the shared library will not be linked against zlib (no linker flag passed). However, `ZLIB_FOUND` in CMake ultimately controls whether the `src/H5Zdeflate.c` file depends on zlib by including `zlib.h`. 

The linker flag will be missing, due to the `ZLIB_LIBRARIES` flag evaluating to an empty string here:
```
 set (LINK_COMP_LIBS ${LINK_COMP_LIBS} ${ZLIB_LIBRARIES})
 ```

I believe what happened on CI is that on some of the docker images, `zlib.h` is installed at the system level (`/usr/include/zlib.h` exists) - and the compiler was falling back to finding it there. The hdf5 shared library was being built (because the linker on Linux does not warn/fail when there are undefined symbols), but then the build was failing further along when linking an executable (the linker does expect all symbols to be defined when linking an executable). I think it's telling that this was failing on the gcc11 build, since that one uses a newer docker image where the compiler is built from source, and zlib is a dependency. I do wonder if some docker images have `zlib.h` and some dont.

Digging further, the `cmake_find_package_multi` may be the culprit here: it generates a `ZLIBConfig.cmake` that will be correctly located by the `find_package(ZLIB)` calls in the hdf5 build scripts - but most importantly, it *does not* set the `ZLIB_LIBRARIES` and `ZLIB_INCLUDE_DIRS` variables (they are suffixed with  `_RELEASE` instead - since this is the "multiconfig" generator). 

So I think it's plausible that we were seing the confluence of 3 different things: 
* the variables not being defined
* the fragility of undefined variables evaluating to empty strings in CMake (using a target would have been far more robust)
* the compiler falling back to `/usr/include` and `zlib.h` happening to be there (it's one of the most common libraries) - potentially only being there on some of the docker images used by Conan Center CI
* the linker not raising issues when linking shared libraries, but doing so when linking executables. 

I believe there should be no issues now by linking directly to `CONAN_PKG::zlib` - however I wonder: why are not linking against `ZLIB::ZLIB` instead? Doing so would be compatible with CMake's built-in find module, _any_ of the current cmake generators from conan 1.x, and also the more modern `CMakeDeps` generator. 


### HDF5 target names
I have found it a bit hard to follow or understand the purpose of this PR with regards to the generated CMake files. My current understanding (and I could be wrong) is that these files set up alias targets so that we have both upper case and lowercase named targets
Recent versions of CMake (>= 3.19)  define the following targets in the  FindHDF5 module:
```
hdf5::hdf5
hdf5::hdf5_hl_cpp
hdf5::hdf5_fortran
hdf5::hdf5_hl
hdf5::hdf5_hl_cpp
hdf5::hdf5_hl_fortran
hdf5::h5diff
```

The HDF5 build itself can also export targets in a generated config, however they follow a slightly different naming convention for the targets, but are also using the `hdf5::` namespace prefix. 

My question would be: if both CMake and the upstream project use a different convention (where all targets and the namespace are lower case), is it not a risk that Conan Center introduces a _third_ convention that does not seem to exist out in the wild?  I can see that lower-case seems to be convention when using targets for hdf5: https://github.com/search?p=1&q=hdf5%3A%3Ahdf5&type=Code - 

I would advocate for aligning ourselves with the existing conventions, no matter how recent they are, and discontinue the use of target names that use the uppercase `HDF5::` prefixes.  

### Warnings from hooks

This warning:
```
post_package_info(): WARN: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] The *.cmake files have to be placed in a folder declared as `cpp_info.builddirs`. Currently folders declared: {'/home/conan/w/prod/BuildSingleReference/.conan/data/hdf5/1.10.5/_/_/package/a3fadebdda8354affd3cc7717ec01e2c13382911/'}
post_package_info(): WARN: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] Found files: ./lib/cmake/conan-official-hdf5-targets.cmake; ./lib/cmake/conan-official-hdf5-variables.cmake
```

I believe this hook checks if the package fines CMake modules that consumers can `include()` simply by refering to them name - this would require the directories that have those files to be in `cpp_info.builddirs`. However, that is not the case here, as the `.cmake` files in question are included directly and explicitly by the generated find modules, so there is no expectations for downstream users to include these files. So I think in this case it would be safe to ignore the warning.

**Comment by paulharris** [comment]
Thanks @jcar87 for the extensive reply,
yes HDF5 cmake files make my head hurt, and unfortunately I'm not even using HDF5 in my own project... I'm just trying to get things to build with conan.  I'm pretty much at the end of my rope here, I can't keep spending time on this.

It would be a lot quicker if someone else who knows more can pick this up and carry it over the line.

On your points:

### cmake_find_package_multi
I used cmake_find_package_multi based on someone's recommendation.  I assume it needs to switch to cmake_find_package instead?   Then there would be no _RELEASE appended.
I'm not even sure it is required at all... how did the recipe work before?  There was no cmake_find_package prior to my PR, and yet it "worked".

### CONAN_PKG
re CONAN_PKG::zlib vs ZLIB::ZLIB ... I did try ZLIB::ZLIB but it didn't work - probably because there is no find_package(ZLIB) call in the patched cmake.   I tried various things and in the end just followed what past contributors had done:
https://github.com/conan-io/conan-center-index/blob/9051a69805f7169353fe738c50106de498e1191e/recipes/hdf5/all/patches/conanize-link-szip-1.12.1%2B.patch#L45
Which was presumably following the pattern set by @SpaceIm back in 2020:
https://github.com/conan-io/conan-center-index/blob/9051a69805f7169353fe738c50106de498e1191e/recipes/hdf5/all/patches/conanize-link-szip-1.8.21.patch#L39

I would use a different pattern if I had known what should be done, but then all the other patches should also be cleaned up, and I don't have the experience or time to do all of that.

### Downstream tests
Might be a good time to make the patch and then feed it back to HDF5 upstream via a PR.
Good test cases I feel are key, and if the HDF5 recipe is updated, it would be better if existing conan packages that depend on HDF5 were also built as part of the CI tests.  Otherwise I have little confidence that the new recipe will not break downstream things.
A good test case would be my VTK package, which is still in draft form.

### Target names

Yes I agree, it would be better if the recipe were consistent with anything that already exists!
Again, this is how I found the recipe, and I don't use it directly so I am not the best person to modify and test it.

### Going forward

It is unclear to me how to proceed with this recipe...
* Just make it work "well enough" and move on?
* Clean up the recipe to use correct patterns?
* Feed back patches to upstream?

I suppose it depends on the conan v2 schedule, because it sounds like there will be a LOT of changes to move all the recipes to the new CMakeDeps patterns.  So perhaps it just needs to be fixed a little further, and then leave notes to the future upgrader.

**Comment by jcar87** [comment]
> Thanks @jcar87 for the extensive reply, yes HDF5 cmake files make my head hurt, and unfortunately I'm not even using HDF5 in my own project... I'm just trying to get things to build with conan. I'm pretty much at the end of my rope here, I can't keep spending time on this.

You're doing a great job! This one was a series of unfortunate events, I'm afraid.

> It would be a lot quicker if someone else who knows more can pick this up and carry it over the line.
> 
> On your points:
> 
> ### cmake_find_package_multi
> I used cmake_find_package_multi based on someone's recommendation. I assume it needs to switch to cmake_find_package instead? Then there would be no _RELEASE appended. I'm not even sure it is required at all... how did the recipe work before? There was no cmake_find_package prior to my PR, and yet it "worked".

The `cmake_find_package_multi` generator is in some ways close to the newer `CMakeDeps`, so that could be a reason why it's recommended. On the other hand, it's main benefit is when doing local development and using a generator that supports switching between build types (Release/Debug etc) - For package building in Conan Center, I don't think the `cmake_find_package_multi` generator would provide advatanges other than being more similar to `CMakeDeps`. I would imagine that with `CMakeDeps` we might have experienced similar problems.

The reason the recipe was working before was that the calls to `find_package(ZLIB)` was falling back to CMake's built-in find module - and it was correctly finding the files from the zlib package inside the conan cache. This module does define the variables that the HDF5 CMake scripts are expecting, so it was all working fine. 

> 
> ### CONAN_PKG
> re CONAN_PKG::zlib vs ZLIB::ZLIB ... I did try ZLIB::ZLIB but it didn't work - probably because there is no find_package(ZLIB) call in the patched cmake. I tried various things and in the end just followed what past contributors had done:
> 
> https://github.com/conan-io/conan-center-index/blob/9051a69805f7169353fe738c50106de498e1191e/recipes/hdf5/all/patches/conanize-link-szip-1.12.1%2B.patch#L45
> 
> 
> Which was presumably following the pattern set by @SpaceIm back in 2020:
> https://github.com/conan-io/conan-center-index/blob/9051a69805f7169353fe738c50106de498e1191e/recipes/hdf5/all/patches/conanize-link-szip-1.8.21.patch#L39
> 
> I would use a different pattern if I had known what should be done, but then all the other patches should also be cleaned up, and I don't have the experience or time to do all of that.

I see! I was not aware of this one, it's generally a great thing to follow existing patterns for consistency :) 

> 
> ### Downstream tests
> Might be a good time to make the patch and then feed it back to HDF5 upstream via a PR. Good test cases I feel are key, and if the HDF5 recipe is updated, it would be better if existing conan packages that depend on HDF5 were also built as part of the CI tests. Otherwise I have little confidence that the new recipe will not break downstream things. A good test case would be my VTK package, which is still in draft form.

I have checked that the minimum version of CMake required by the HDF5 CMakeLists already has support for the `ZLIB::ZLIB` target as provided by the FindZLIB module. So I think there is a potential upstream contribution. If ZLIB::ZLIB had been used, it would have worked with more cmake generators from conan (as long as the zlib recipes expresses that the target and the cmake package have the names in upper case), but most importantly, CMake would fail generating if any target was linking agsint `ZLIB::ZLIB` but nobody was defining it. This is more robust than the variables silently evaluating to an empty string, and then the compiler falling back to finding the required headers, by chance, in the system path.
> 
> ### Target names
> Yes I agree, it would be better if the recipe were consistent with anything that already exists! Again, this is how I found the recipe, and I don't use it directly so I am not the best person to modify and test it.

I have seen other recipes in Conan center where this happens - that is, using CMake package or target names that do not exist out in the wild. This is a consequence of how packages and targets have evolved in CMake, and the current lack of consistent convention.


> ### Going forward
> It is unclear to me how to proceed with this recipe...
> 
> * Just make it work "well enough" and move on?

I think the recipe as it currently is in this PR does the right thing. Is it still the case that it's not passing CI properly?

> * Clean up the recipe to use correct patterns?
> * Feed back patches to upstream?

Down the line, both :) I'd be happy to help!

> 
> I suppose it depends on the conan v2 schedule, because it sounds like there will be a LOT of changes to move all the recipes to the new CMakeDeps patterns. So perhaps it just needs to be fixed a little further, and then leave notes to the future upgrader.

**Comment by paulharris** [comment]
I'm confused, the CI now builds this?
* I haven't changed cmake_find_package_multi to cmake_find_package ... not a problem now?
* Still uses CONAN_PKG:: ... but this is ok until conan-v2 ?

Did the CI change?

## Upstream HDF5 changes to simplify the zlib + szip
I might be able to squeeze in a patch for upstream,
I wonder if all of that boils down to a find_package(ZLIB) and `set (LINK_COMP_LIBS ${LINK_COMP_LIBS} ZLIB::ZLIB)`
We can test that here on another draft PR, so we can at least get this PR merged first?
